Providers/Scripts/2.4x-2.5x/Scripts/nxOMSAgentNPMConfig.py

```diff
@@ -274,67 +300,175 @@ def TestConfigUpdate(Contents):
+        for x in wsListFromState:
+            if x not in wsToRemove:
+                agentOwningWsChanged = True
+                wsFormerRunningNPM = wsRunningNPM
+                wsRunningNPM = x
+                break
+
+    # Check if purge should be called
+    if (not wsListFromState) or len(wsListFromState) == len(wsToRemove):
+        LOG_ACTION.log(LogType.Info, 'Purging solution as no workspaces should be running NPM')
+        retval = PurgeSolution()
+    else:
+        retval = UpdateStateFile(wsListFromState, wsToRemove, wsRunningNPM)
+        if (not retval) and agentOwningWsChanged:
+            if wsFormerRunningNPM is not None:
+                LG().Log(LogType.Info, 'Restarting omsagent of previous workspace ' + wsFormerRunningNPM)
+                retval = OMS_ACTION.restart_oms_agent(wsFormerRunningNPM)
+            LG().Log(LogType.Info, 'Restarting omsagent of workspace ' + wsRunningNPM)
+            retval &= UpdatePluginConfFiles(wsRunningNPM)
+
+    return retval
+
+def SetConfigUpdate(ConfigID, Contents):
+    destFileDirPath = os.path.join(PREFIX_ETC, ConfigID, RPATH_AGENT_CONF)
+    destFileFullPath = os.path.join(destFileDirPath, FNAME_AGENT_CONF)
 
     # Update config after checking if directory exists
-    if not os.path.exists(CONFIG_PATH):
-        LOG_ACTION.log(LogType.Error, 'CONFIG_PATH does not exist')
+    if not os.path.exists(destFileDirPath):
+        LG().Log(LogType.Error, 'Conf directory for workspace ' + ConfigID + ' does not exist')
         retval = -1
     else:
         retval = WriteFile(destFileFullPath, Contents)
-        if retval == 0 and os.path.exists(AGENT_RESOURCE_VERSION_PATH): #notify server only if plugin is present
-            LG().Log(LogType.Info, 'Updated the file, going to notify server')
+        LOG_ACTION.log(LogType.Info, 'Updated configuration file for ' + ConfigID)
+        # Notify NPM of config change if single workspace scenario
+        if (not retval) and len(GetWorkspaceListFromStateFile()) == 1:
             retval = NotifyServer(Commands.Config)
     return retval
 
+# Needs to run after state file updation in set
 def SetFilesUpdate(newVersion):
+    wsRunningNPM = GetWorkspaceRunningNPMAgent()
     retval = UpdateAgentBinary(newVersion)
     retval &= UpdatePluginFiles()
+    if wsRunningNPM is not None:
+        retval &= UpdatePluginConfFiles(wsRunningNPM)
     if retval:
         return 0
     return -1
 
+def UpdateStateFile(completeList, removeList, owner):
+    retval = 0
+    if len(completeList) != 0 and (owner is not None) and (owner in completeList):
+        listToWrite = []
+        listToWrite.append(owner)
+        for x in completeList:
+            if x != owner and x not in removeList:
+                listToWrite.append(x)
+        contents = ''
+        for x in listToWrite:
+            contents += x
+            contents += '\n'
+        retval = WriteFile(os.path.join(PATH_NPM_STATE, FNAME_NPM_STATE), contents)
```

Comment:
Is this directory guaranteed to exist on all paths where `UpdateStateFile` is called?